File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/Python/optimal_merge_pattern.py
import heapq 
from typing import List

from sympy import sec 

class OptimalMergePattern:
    def __init__(self,file_sizes:List[int]):
        self.file_sizes=file_sizes.copy() # Set of files with diff. sizes 

    def minimum_cost(self)->int:
        if(len(self.file_sizes)<2):
            return 0
        min_heap = self.file_sizes.copy()
        heapq.heapify(min_heap)

        cost = 0 
        while len(min_heap)>1:
            first_smallest = heapq.heappop(min_heap)
            second_smallest = heapq.heappop(min_heap)
            merged_size = first_smallest + second_smallest
            cost+=merged_size
            heapq.heappush(min_heap,merged_size)
        return cost 
    
def main():
    file_sizes = [10,55,32,8,24]
    optimal_cost_finder = OptimalMergePattern(file_sizes)
    print(f"Set of Files with Sizes : {file_sizes} \n Optimal minimum merge cost : {optimal_cost_finder.minimum_cost()}")

if __name__ =="__main__":
    main()

File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/C++/optimal_merge_pattern.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std ; 

class OptimalMergePattern{
vector<int>file_sizes; 
public: 
explicit OptimalMergePattern(const vector<int>&files): file_sizes(files){}
int minimum_cost(){
    if(file_sizes.size()<2)return 0;
    priority_queue<int,vector<int>,greater<int>>min_heap(file_sizes.begin(),file_sizes.end());
    int cost = 0 ;
    while(min_heap.size()>1){
        int first_smallest = min_heap.top(); min_heap.pop(); 
        int second_smallest = min_heap.top(); min_heap.pop(); 
        int merged_size= first_smallest+second_smallest; 
        cost+=merged_size;
        min_heap.push(merged_size); 
    }

    return cost; 
}
};

int main(){
    vector<int>file_sizes = {10,55,32,8,24}; 
    OptimalMergePattern optimal_cost_finder(file_sizes); 
    cout<<"Set of File Sizes : "; 
    for(int size : file_sizes)cout<<size<<" "; cout << endl ; 
    cout<<"Minimum optimal Merge Cost : "<<optimal_cost_finder.minimum_cost()<<endl ; 
}